### PR TITLE
Ticket30914

### PR DESCRIPTION
**Pull request by nmathewson**

### CODE REVIEW

File: src/app/config/config.c [hunk 1]
```diff
@@ -253,23 +254,45 @@ static config_abbrev_t option_abbrevs_[] = {
  * members with CONF_CHECK_VAR_TYPE. */
 DUMMY_TYPECHECK_INSTANCE(or_options_t);
 
-/** An entry for config_vars: "The option <b>name</b> has type
+/** An entry for config_vars: "The option <b>varname</b> has type
```
Comment (teor2345) [inline src/app/config/config.c]: Please document macro argument initvalue.

File: src/app/config/statefile.c [hunk 2]
```diff
@@ -71,8 +71,10 @@ static config_abbrev_t state_abbrevs_[] = {
 DUMMY_TYPECHECK_INSTANCE(or_state_t);
 
 /*XXXX these next two are duplicates or near-duplicates from config.c */
```
Comment (teor2345) [inline src/app/config/statefile.c]: Will these duplicates will be refactored in some future PR?

Same question for shared random and other modules.

File: src/lib/conf/conftypes.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                             * mixed with other keywords. */
+  CONFIG_TYPE_LINELIST_V,   /**< Catch-all "virtual" option to summarize
+                             * context-sensitive config lines when fetching.
+                             */
+  // XXXX this doesn't belong at this level of abstraction.
+  CONFIG_TYPE_ROUTERSET,    /**< A list of router names, addrs, and fps,
+                             * parsed into a routerset_t. */
+  CONFIG_TYPE_OBSOLETE,     /**< Obsolete (ignored) option. */
+  CONFIG_TYPE_EXTENDED,     /**< Extended type; definition will appear in
+                             * pointer. */
```
Comment (teor2345) [inline src/lib/conf/conftypes.h]: Are extended types also going to become part of a higher level of abstraction?

File: src/lib/confmgt/type_defs.c [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+
+static int
+double_parse(void *target, const char *value, char **errmsg,
+             const void *params)
+{
+  (void)params;
+  (void)errmsg;
+  double *v = (double*)target;
+  // XXXX This is the preexisting behavior, but we should detect errors here.
+  *v = atof(value);
```
Comment (teor2345) [inline src/lib/confmgt/type_defs.c]: When are we going to detect errors here?

File: src/test/test_confparse.c [hunk 5]
```diff
@@ -46,15 +48,17 @@ typedef struct test_struct_t {
 
 static test_struct_t test_struct_t_dummy;
 
-#define VAR(name,conftype,member,initvalue)                             \
-  { name, CONFIG_TYPE_##conftype, offsetof(test_struct_t, member),      \
+#define VAR(varname,conftype,member,initvalue)                          \
```
Comment (teor2345) [inline src/test/test_confparse.c]: These are also duplicates, which should get removed at some future time.
(Unless they are necessary for low-level tests. If they are, they should get a comment saying so.)

File: src/test/test_confparse.c [hunk 6]
```diff
@@ -805,6 +832,77 @@ test_confparse_extra_lines(void *arg)
...
+  tt_assert(ok);
+
+  /* Try a couple of unitless values. */
+  tt_int_op(config_parse_interval("10", &ok), OP_EQ, 10);
+  tt_assert(ok);
+  tt_u64_op(config_parse_interval("15.0", &ok), OP_EQ, 15);
+  tt_assert(ok);
+
+  /* u64 overflow */
+  /* XXXX our implementation does not currently detect this. See bug 30920. */
```
Comment (teor2345) [inline src/test/test_confparse.c]: Are we going to fix #30920 as part of this refactor?
Preserving existing behaviour is also ok.

File: src/test/test_confparse.c [hunk 7]
```diff
@@ -805,6 +832,77 @@ test_confparse_extra_lines(void *arg)
...
+  /* i32 overflow */
+  tt_int_op(config_parse_interval("1000 months", &ok), OP_EQ, -1);
+  tt_assert(!ok);
+  tt_int_op(config_parse_msec_interval("4 weeks", &ok), OP_EQ, -1);
+  tt_assert(!ok);
+
+  /* bad units */
+  tt_u64_op(config_parse_memunit("7 nybbles", &ok), OP_EQ, 0);
+  tt_assert(!ok);
+  // XXXX these next two should return -1 according to the documentation.
```
Comment (teor2345) [inline src/test/test_confparse.c]: Is this comment still true?

File: src/app/config/config.c (hunk 1 above)
Comment (nmathewson) [reply src/app/config/config.c]: This happens later in the #29211 branch, in c553750e32d1bf669a3e8308fa44319954a627ca when we move all these macros into src/lib/confmacros.h . I believe that is part of #30935.

File: src/app/config/statefile.c (hunk 2 above)
Comment (nmathewson) [reply src/app/config/statefile.c]: Yes: also part of #30935.

File: src/lib/conf/conftypes.h (hunk 3 above)
Comment (nmathewson) [reply src/lib/conf/conftypes.h]: I think so, if I understand your question -- see the definition of struct_member_t.  The first extended type is routerset, which is first implemented in e16b90b88a76.

File: src/lib/confmgt/type_defs.c (hunk 4 above)
Comment (nmathewson) [reply src/lib/confmgt/type_defs.c]: I have opened a ticket as https://trac.torproject.org/projects/tor/ticket/31475#ticket

File: src/test/test_confparse.c (hunk 5 above)
Comment (nmathewson) [reply src/test/test_confparse.c]: This also gets fixed when we combine everything into serc/lib/confmacros.h in c553750 as a part of #30935.

File: src/test/test_confparse.c (hunk 6 above)
Comment (nmathewson) [reply src/test/test_confparse.c]: I think it should get fixed independently, after this code is stable.

File: src/test/test_confparse.c (hunk 7 above)
Comment (nmathewson) [reply src/test/test_confparse.c]: I believe it is, yes.  The documentation says that if the intervals are bad it should return -1 IIC

Comment (teor2345) [reply src/test/test_confparse.c]: Ok so let's update the documentation for parse_interval() to document the special cases.

The doc change might belong in another ticket.

**Pull request merged**